lib/types/host-rules.ts

```diff
@@ -1,4 +1,4 @@
-export interface HostRuleSearchResult {
+export interface HostRuleSearchResult extends Record<string, any> {
```

Comment:
Avoid using any

Reply:
what would you suggest instead of `any` here? would `unknown` works?

Reply:
the reason behind this line change here is tsc will complain on the `rule[suffix]` references in [here](https://github.com/lyonlai/renovate/blob/main/lib/workers/global/config/parse/host-rules-from-env.ts#L31)

```
lib/workers/global/config/parse/host-rules-from-env.ts:61:9 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'HostRule'.
  No index signature with a parameter of type 'string' was found on type 'HostRule'.

61         rule[suffix] = env[envName];
```

would be appreciated for any better way forward.

Reply:
```diff
diff --git a/lib/modules/datasource/packagist/index.spec.ts b/lib/modules/datasource/packagist/index.spec.ts
index e445de179..a4a02f7a8 100644
--- a/lib/modules/datasource/packagist/index.spec.ts
+++ b/lib/modules/datasource/packagist/index.spec.ts
@@ -2,7 +2,8 @@ import { mockDeep } from 'jest-mock-extended';
 import { getPkgReleases } from '..';
 import { Fixtures } from '../../../../test/fixtures';
 import * as httpMock from '../../../../test/http-mock';
-import type { HostRule } from '../../../types';
+import type { HostRuleSearchResult } from '../../../types';
+import type { HostRuleSearch } from '../../../util/host-rules';
 import * as _hostRules from '../../../util/host-rules';
 import * as composerVersioning from '../../versioning/composer';
 import { id as versioning } from '../../versioning/loose';
@@ -25,7 +26,7 @@ describe('modules/datasource/packagist/index', () => {
     let config: any;

     beforeEach(() => {
-      hostRules.find = jest.fn((input: HostRule) => input);
+      hostRules.find = jest.fn((input: HostRuleSearch) => input as HostRuleSearchResult);
       hostRules.hosts = jest.fn(() => []);
       config = {
         versioning: composerVersioning.id,
diff --git a/lib/types/host-rules.ts b/lib/types/host-rules.ts
index 6206b9c81..bb9d689aa 100644
--- a/lib/types/host-rules.ts
+++ b/lib/types/host-rules.ts
@@ -1,4 +1,4 @@
-export interface HostRuleSearchResult extends Record<string, any> {
+export interface HostRuleSearchResult extends Record<string, unknown> {
   authType?: string;
   token?: string;
   username?: string;
```

I've tried locally with `unknown`. There were some tiny issue with test mocking in datasource. Here is the diff. Let me know if it is better. @rarkins

Reply:
@viceice @zharinov would some "key of" TS trickery work here?

Reply:
I have tried to use `keyof` tricks on the object index
both `rule[suffix as keyof typeof rule] = env[envName]` or `rule[suffix as keyof HostRule] = env[envName]` gives me `Type 'string | undefined' is not assignable to type undefined`

Reply:
we can't simply copy that values, as some of them need to be numbers. so you need a switch case. that will also solve the typescript issue.

Reply:
that's a good one. thank you. I've wrapped it up in a function